File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/video/RxVideoPlayer.java
package com.pr0gramm.app.ui.views.viewer.video;

import android.support.annotation.Nullable;

import rx.Observable;
import rx.subjects.BehaviorSubject;
import rx.subjects.PublishSubject;

/**
 * This is a base for {@link VideoPlayer}s with some rx features.
 */
public abstract class RxVideoPlayer implements VideoPlayer {
    protected final PublishSubject<Void> detaches = PublishSubject.create();
    final BehaviorSubject<Boolean> buffering = BehaviorSubject.create(false);
    Callbacks clientCallbacks = new EmptyVideoCallbacks();

    @Override
    public void setVideoCallbacks(@Nullable Callbacks callbacks) {
        this.clientCallbacks = callbacks != null ? callbacks : new EmptyVideoCallbacks();
    }

    public Observable<Boolean> buffering() {
        return buffering;
    }

    public Observable<Void> detaches() {
        return detaches;
    }

    protected final Callbacks callbacks = new Callbacks() {
        @Override
        public void onVideoBufferingStarts() {
            clientCallbacks.onVideoBufferingStarts();
            buffering.onNext(true);
        }

        @Override
        public void onVideoBufferingEnds() {
            clientCallbacks.onVideoBufferingEnds();
            buffering.onNext(false);
        }

        @Override
        public void onVideoRenderingStarts() {
            onVideoBufferingEnds();
            clientCallbacks.onVideoRenderingStarts();
        }

        @Override
        public void onVideoSizeChanged(int width, int height) {
            clientCallbacks.onVideoSizeChanged(width, height);
        }

        @Override
        public void onVideoError(String message, ErrorKind kind) {
            clientCallbacks.onVideoError(message, kind);
        }

        @Override
        public void onDroppedFrames(int count) {
            clientCallbacks.onDroppedFrames(count);
        }
    };
}

File: app/src/main/java/com/pr0gramm/app/ui/views/viewer/video/RxVideoPlayer.kt
package com.pr0gramm.app.ui.views.viewer.video

import rx.Observable
import rx.subjects.BehaviorSubject

/**
 * This is a base for [VideoPlayer]s with some rx features.
 */
abstract class RxVideoPlayer : VideoPlayer {
    protected val detaches: BehaviorSubject<Unit> = BehaviorSubject.create()

    override var videoCallbacks: VideoPlayer.Callbacks? = null

    private val buffering = BehaviorSubject.create<Boolean>(false)

    fun buffering(): Observable<Boolean> {
        return buffering
    }

    fun detaches(): Observable<Unit> {
        return detaches
    }

    protected val callbacks: VideoPlayer.Callbacks = object : VideoPlayer.Callbacks {
        override fun onVideoBufferingStarts() {
            videoCallbacks?.onVideoBufferingStarts()
            buffering.onNext(true)
        }

        override fun onVideoBufferingEnds() {
            videoCallbacks?.onVideoBufferingEnds()
            buffering.onNext(false)
        }

        override fun onVideoRenderingStarts() {
            onVideoBufferingEnds()
            videoCallbacks?.onVideoRenderingStarts()
        }

        override fun onVideoSizeChanged(width: Int, height: Int) {
            videoCallbacks?.onVideoSizeChanged(width, height)
        }

        override fun onVideoError(message: String, kind: VideoPlayer.ErrorKind) {
            videoCallbacks?.onVideoError(message, kind)
        }
    }
}
